---
title: "scDown::run_scvelo - Incorporate spliced and unspliced counts and conduct RNA velocity analysis"
output: 
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    toc_depth: 3        # Set the depth of the outline (1 for main headers, 2 for subheaders, etc.)
    number_sections: true  # Automatically number sections
    #theme: united  # many options for theme, this one is my favorite.
#bibliography: library.bib
fig_width: 8 
fig_height: 5 
vignette: >
  %\VignetteIndexEntry{scVelo-1-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = TRUE,
  eval = FALSE,     # Ensure code is not run by default
  message = FALSE,
  warning = FALSE,
  fig.width=8, 
  fig.height=5
)
```

```{r setup}
```

# Define Variables specific for `scDown::run_scvelo`
Apart from global universal variables applied to all functions in scDown, defining additional variables specific for running `scDown::run_scvelo`. 

Loading a Seurat object (seurat_obj) containing single-cell RNA-seq data is required. The cell type annotation column (annotation_column) is defined to specify the cell type or other annotation labels, ensuring the analysis runs with the correct group.

The `loom_files` is required for `run_scvelo`, except when the spliced and unspliced mRNA counts are already incoporated into the Seurat object. The loom files containing spliced and unspliced counts can be generated using [`velocyto`](https://velocyto.org/velocyto.py/tutorial/index.html). Please refer to [`velocyto`](https://velocyto.org/velocyto.py/install/index.html) for instalation and detailed usage instructions. Below is sample code to generate loom files for each sample using [`velocyto`](https://velocyto.org/velocyto.py/tutorial/cli.html#run10x-run-on-10x-chromium-samples): 
```{sh, message = FALSE}
# run velocyto on 10X Chromium samples 
velocyto run10x -v /path/to/cellranger/output/sample genes.gtf 

# /path/to/cellranger/output/sample: CellRanger output folder containing the subfolder outs/, with the 
# necessary BAM files: outs/possorted_genome_bam.bam and outs/possorted_genome_bam.bam.bai

# gene.gtf needs to correspond to correct species, such as human and mouse:
# - refdata-gex-GRCh38-2020-A/genes/genes.gtf (for human)
# - refdata-gex-mm10-2020-A/genes/genes.gtf (for mouse)
```

To integrate spliced and unspliced mRNA counts from loom files with the Seurat object, users need to specify `loom_file_subset_column` to indicate the metadata column in the seurat object that matches with the file names of input loom files (the default is "orig.ident"). For multiple loom files, the file names in `loom_files` must correspond to the values in the specified metadata column of the Seurat object.

By default, `loom_file_subset_by` is left blank, as it will automatically be extracted from loom file names in the same order as input `loom_files`. If needed, users may specify a character vector for `loom_file_subset_by` to subset the Seurat object while ensuring that the order of `loom_file_subset_by` matches the order of file names in `loom_files`. 

The mode for scVelo velocity calculation can be one of four options: "steady_state" (original), "deterministic", "stochastic" (fastest: recommended, default), "dynamical". 

The working directory `output_dir` by default is the current directory and can be changed to specific path. 


```{r, message = FALSE}
# file names for input loom files containing spliced and unspliced counts of scRNA-seq data
# e.g., loom_files=c("/path/to/directory/WT.loom", "/path/to/directory/KO.loom") 
loom_files=NULL
# specify the metadata column in the Seurat object that matches with the file names of input loom_files
loom_file_subset_column="orig.ident"
# optional, this can automatically be extracted from file names of input loom_files 
loom_file_subset_by=NULL

# mode for scVelo velocity calculation
mode='stochastic'

# the number of grids along each axis, controlling the number of vectors on umap
grid_resolutions=c(50)
# velocity vector size (arrow head size)
arrow_sizes=c(0.5,1)
# velocity vector size (vector width)
vector_widths=0.5

# colors to be used in plotting
color_scale=NULL
# specify which metadata column of the Seurat object is the color scale named by
name_by=NULL

```

# Run `scDown::run_scvelo`: Read loom using velocyto.R and run scVelo using the R wrapper velociraptor
The spliced and unspliced matrices from loom files are added as new assays to seurat_obj, and then converted to h5ad file, which will be used as input for `run_scvelo_full`. 

## Run scVelo for entire data
We estimate RNA velocity for entire data using spliced and unspliced counts of scRNA-seq data using scVelo and visualize the vector field. 

```{r}
start_time <- proc.time()

# Run scVelo Part I
run_scvelo(seurat_obj = seurat_obj,
           loom_files = loom_files,   
           loom_file_subset_column = "orig.ident",       
           output_dir = output_dir,
           annotation_column = annotation_column )

end_time <- proc.time()
elapsed_time <- end_time - start_time

# Print the elapsed time
print(elapsed_time)

```
```{r}
# elapsed_time: 1 min
```
This plot visualizes the vector field on UMAP embeddings using different arrow head sizes and widths. 

```{r, echo = FALSE, eval = TRUE}

knitr::include_graphics("../scdown/scvelo/velocityField__gridRes50_arrowSize0.5_width0.5.png")

```

```{r, echo = FALSE, eval = TRUE}

knitr::include_graphics("../scdown/scvelo/velocityField__gridRes50_arrowSize1_width0.5.png")

```



## Run scVelo for specific groups 
If specific groups of conditions or time points are provided, we subset the Seurat object for each group of conditions or time points, and then compute velocity using scVelo and visualize. 

```{r}
# subset a group of conditions or time points used to calculate RNA velocity
groups=list("12","35")
# specify which metadata column of the Seurat object should be used for the group subset 
group_column="age.days."
# specify the number of cores to run each group separately in parallel
cores = 2

# Run scVelo Part I
run_scvelo(seurat_obj = seurat_obj,
           loom_files = loom_files,   
           loom_file_subset_column = "orig.ident",  
           output_dir = output_dir,
           annotation_column = annotation_column,
           groups = groups,
           group_column = group_column,
           cores = cores )

```

This plot visualizes the vector field on UMAP embeddings for each time point group. 

```{r}
# time point: day 12 
```
```{r, echo = FALSE, eval = TRUE}

knitr::include_graphics("../scdown/scvelo/velocityField_12_gridRes50_arrowSize0.5_width0.5.png")

```
```{r}
# time point: day 35 
```
```{r, echo = FALSE, eval = TRUE}
knitr::include_graphics("../scdown/scvelo/velocityField_35_gridRes50_arrowSize0.5_width0.5.png")

```

# Summary

This document outlines a pipeline for incorporating spliced and unspliced RNA matrices from .loom files into Seurat object, and analyzing RNA velocity using velciraptor, a R wrapper for scVelo. The pipeline begins by defining key variables, such as the working directory, Seurat object, species, and metadata for cell type annotations and generate h5ad file as an input for the function `scDown::run_scvelo_full`. It runs scVelo on the entire dataset to compute RNA velocity across all cells and generates generates visualizations for Velocity vector field on UMAPs.

Additionally, the pipeline supports focused analysis by selecting specific time points  (e.g., age groups 12 and 35) or comparing user-defined groups. For each group of conditions or time points, the Seurat object is subsetted and velocity is computed in parallel. 

This pipeline provides a comprehensive tool for incorporating spliced and unspliced matrices from loom file into the Seurat object and computing RNA velocity.

The vignette was performed on a machine with the following specifications:

- **Operating System**: Rocky Linux 8.9 (Green Obsidian)
  - **Product Name**: RHEL
  - **Product Version**: 8.9
- **CPU**: Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
- **Memory**: 10 GB