---
title: "scDown::run_monocle3 - Construction of the pseudotime trajectories to model the progression of cellular differentiation"
author: |
    | 
#date: "`r format(Sys.Date(), '%m/%Y')`"
output: 
  rmarkdown::html_document:
    highlight: pygments
    toc: true
    toc_depth: 3
    number_sections: true  ## if you want number sections at each table header
    #theme: united  # many options for theme, this one is my favorite.
#bibliography: library.bib
fig_width: 8 
fig_height: 5 
vignette: >
  %\VignetteIndexEntry{scDown_moncole}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  fig.width=8, fig.height=5
)

```

# Define variables specific for run_monocle3

```{r, message = FALSE,eval=FALSE}
# PCA Dimensions to preprocess the data
nDim <- 30 

```

# Run Monocle3

## Run monocle3 default workflow

Users can load their scRNASeq data which they would like to run the trajectory analysis. To run the trajectory analysis with `run_monocle3()`, the users have to provide the scRNASeq data and species. By default, the function runs the trajectory analysis on the whole dataset, identify the root nodes using potency method, and run trajectory analysis using monocle3. While `run_monocle3()` is generally species-agnostic and works beyond mouse or human, species specification is required when using `rootNodes_method = "potency"`. If you choose to manually define root nodes using `rootNodes_method = "rootNodes"`, then species specification is not necessary.

```{r, message = FALSE,eval=FALSE}
run_monocle3(seurat_obj = seurat_obj,
             species = species,
             nDim = nDim,
             annotation_column = annotation_column,
             graph_test=TRUE,
             output_dir = output_dir)

```

**Running Time**: ~ 5 minutes

**Operating System**: Rocky Linux 8.9 (Green Obsidian)

  - **Product Name**: RHEL
  - **Product Version**: 8.9
  
**CPU**: Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz

**Memory**: 10 GB

### Pre-processing and visualization of the single cell data
The first step of construction of the single cell trajectories is the pre-processing step of the data. In this step, the data is normalized and batch corrected (if the user provides the batch metadata in the `batch_metadata` variable. By Default, it is NULL). After that, the dimensionality reduction is carried out and UMAP coordinates are calculated. The function can also use the UMAP coordinates from the Seurat object (by setting `transferUMAP=TRUE`). The UMAP plot shows the different cell types in the data after transferring the UMAP coordinates.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/umap_celltypes_transferUMAP_TRUE.png")

```

Next, the cells are grouped into broad cell clusters represented in the data. Monocle does not assume that all cells in the dataset descend from a common transcriptional "ancestor". In many experiments, there might in fact be multiple distinct trajectories. The monocle functions uses a technique called community detection to group cells. This approach was introduced by Levine et al as part of the phenoGraph algorithm. The UMAP plot shows the different partitions in the data.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/umap_partitions_transferUMAP_TRUE.png")

```

### Construction and visualization of single-cell trajectories
Next, the trajectories in the single-cell data is learned by fitting a principal graph within each partition. These trajectories are further used to fit or order the cells in the pseudotime. Pseudotime is a measure of how much progress an individual cell has made through a process such as cell differentiation. In order to do that, monocle needs a root node which acts as a starting point for the differentiation or the biological process. The UMAP plot shows the different nodes along the trajectories.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/umap_trajectory_transferUMAP_TRUE.png")

```

In our workflow, we have three methods to identify the root node in the single cell data. By default (`rootNode_method = "potency"`), we use the methods developed by Teschendorff et al. which uses the their gene expression profiles and a protein-protein interaction network to automatically identify the root node based on the potency. Below we show the UMAP plot that ordered the cells based on the root node selected using potency.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/umap_pseudotime_potency.png")

```

The second method is to automatically identify the root node which is most heavily occupied by early time point cells ((`rootNode_method = "rootNodes"`)). For that, the users need to provide the early timepoint category (`timepoint`) and the meta data column (`timePoint_metadata`) in the seurat object. The third method is that users can provide the root name by selecting and passing the principal point (`rootNode`) on the trajectory in the trajectory UMAP plot.  

### Identifiation of significant trajectory-variable genes and modules
The pipeline also identifies the genes that vary between groups of cells across the trajectories using the spatial autocorrelation analysis called Moran's I developed by Cao & Spielmann et al. This method identifies genes that vary between groups of cells in UMAP space and is effective in finding genes that vary in single-cell RNA-seq datasets. All the significant trajectory-variable genes were exported in a csv file (`monocleDEG_significant_by_trajectory.csv`).The plot below shows the expression of the top trajectory-variable genes in a feature plot.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/significant_by_trajectory_gene_featurePlot.png")

```

We also visualized the expression levels of top trajectory-variable genes across the pseudotime and colored as cell types.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/significant_by_trajectory_genesInPseudotime.png")

```

The set of the trajectory-variable genes is further grouped into modules using Louvain community analysis. The clustering information is exported in a csv file (`monocleDEG_by_trajectory_moduleInformation.csv`). The module heatmap below shows the expression levels of each module across the cell types.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/significant_by_trajectory_moduleHeatmap.png")

```

### Visualization of cell distribution across the pseudotime
The function will also generate plots that shows the distribution of the cells across the pseudotime. Below the density plot shows the distribution of cells across the pseudotime for the complete dataset separated by the timepoints.

```{r, echo = FALSE, eval = TRUE, out.width = "60%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/cell_distribution_density-NoCond.png")

```

Below the histogram shows the distribution of cells across the pseudotime for the complete dataset per cell type separated by timepoints.

```{r, echo = FALSE, eval = TRUE, out.width = "60%", out.height = "80%"}

knitr::include_graphics("../scdown/monocle3/images/celltype_distribution_histogram-NoCond.png")

```

The workflow also generates the distribution of cells per condition too in the output folder.

## Run monocle3 with user defined groups
In this part, we will show how the pipeline can be used to analyze the trajectories between the different conditions and compare the expression of the significant trajectory variable genes across the conditions using regression analysis.

### Construction and visualization of single-cell trajectories per group
The users need to provide the conditions and condition metadata in the `run_monocle3` function to carry out the trajectory analysis across different conditions. This workflow also need the `deg_method` (By default, `deg_method = "quasipoisson"`) to carry out the regression analysis to identify those trajectory variable genes that differ between the conditions. In our test data, we selected the age of the mouse in days (12 and 35 days) as the conditions. 

```{r, message = FALSE,eval=FALSE}
conditions <- c("12","35") # a string specifying a column in seurat_object@meta.data that contains the condition names, or NULL when no condition is filled.
condition_metadata <- "age.days."
deg_method = "quasipoisson"

library(scDown)
run_monocle3(seurat_obj=seurat_obj,
             species=species,
             nDim = nDim,
             annotation_column = annotation_column,
             graph_test=TRUE,
             conditions = conditions,
             group_column=condition_metadata,
             deg_method = "quasipoisson",
             output_dir = output_dir)

```
**Running Time**: ~ 8 minutes

**Operating System**: Rocky Linux 8.9 (Green Obsidian)

  - **Product Name**: RHEL
  - **Product Version**: 8.9
  
**CPU**: Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz

**Memory**: 10 GB

The UMAP plot shows the different cell types in the data at timepoint 12.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/umap_celltypes_transferUMAP_TRUETRUE_12.png")

```

Below we show the UMAP plot that ordered the cells based on the root node selected using potency at timepoint 12.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/umap_pseudotime_potency_12.png")

```

### Visualization of cell distribution across the pseudotime separated by conditions
When conditions are supplied, the function will also generate plots that shows the distribution of the cells across the pseudotime separated by conditions. Below the density plot shows the distribution of cells across the pseudotime for the complete dataset separated by the timepoints.

```{r, echo = FALSE, eval = TRUE, out.width = "60%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/cell_distribution_density.png")

```

Below the histogram shows the distribution of cells across the pseudotime for the complete dataset per cell type separated by timepoints.

```{r, echo = FALSE, eval = TRUE, out.width = "60%", out.height = "80%"}

knitr::include_graphics("../scdown/monocle3/images/celltype_distribution_histogram.png")

```

The workflow also generates the distribution of cells per condition too in the output folder.


### Identifiation of significant differential trajectory-variable genes between conditions
When conditions are supplied, the pipeline also identifies the significant differential trajectory-variable genes between conditions by fitting a regression model (using the conditions metadata) to each identified trajectory-variable genes. All the significant differential trajectory-variable genes were exported in a csv file (`monocleDEG_significant_by_age.days.+trajectory_12_35.csv`).The plot below shows the expression of the top trajectory-variable genes in a feature plot.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/significant_by_age.days.+trajectory_12_35_featurePlot.png")

```

We also visualized the expression levels of top trajectory-variable genes across the pseudotime and colored as cell types.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/significant_by_age.days.+trajectory_12_35_genesInPseudotime.png")

```


### Regression analysis between various groups in the scRNASeq data
The pipeline can also be used to identify the significant differential expression genes between conditions using a generalized linear model using "quasipoisson" distribution by default. The users can use four different models ("quasipoisson", "negbinomial","poisson" or "binomial") by supplying it in `deg_model` variable. This is an optional feature which can be turned on by supplying the metadata in the `metadata_deg_model`.

```{r, message = FALSE,eval=FALSE}
conditions <- c("12","35") # a string specifying a column in seurat_object@meta.data that contains the condition names, or NULL when no condition is filled.
condition_metadata <- "age.days."
deg_method = "quasipoisson"

library(scDown)
run_monocle3(seurat_obj=seurat_obj,
             species=species,
             nDim = nDim,
             annotation_column = annotation_column,
             deg_method = "quasipoisson",
             metadata_deg_model=condition_metadata,
             output_dir = output_dir)

```
**Running Time**: ~ 8 minutes

All the significant differential expressed genes were exported in a csv file (`monocleDEG_significant_by_age.days..csv`).The plot below shows the expression of the top differential expressed genes in a feature plot.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/significant_by_age.days._featurePlot.png")

```

We also visualized the expression levels of top differential expressed genes in violin plot.
```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/significant_by_age.days._violinPlot.png")

```


## Run monocle3 with user defined cell types groups

In this workflow, the users can also provide list of combinations of cell types between which they would like to analyze using the trajectory analysis workflow.

```{r, message = FALSE,eval=FALSE}
library(scDown)

cell_types_monocle<-list(c("Granule immature","Granule mature"))
library(scDown)
run_monocle3(seurat_obj=seurat_obj,
             species=species,
             nDim = nDim,
             annotation_column = annotation_column, 
             conditions = conditions,
             group_column=condition_metadata,
             deg_method = "quasipoisson",
             celltype_groups = cell_types_monocle,
             graph_test=TRUE,
             output_dir = output_dir)

```
**Running Time**: ~10 minutes

**Operating System**: Rocky Linux 8.9 (Green Obsidian)

  - **Product Name**: RHEL
  - **Product Version**: 8.9
  
**CPU**: Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz

**Memory**: 10 GB

### Construction and visualization of single-cell trajectories
As mentioned previously, the trajectory in the cell type groups is learned by fitting a principal graph within each partition which are further used to fit or order the cells in the pseudotime. The UMAP plot shows the different nodes along the trajectories for the cell type groups.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/umap_trajectory_transferUMAP_TRUE_Granule immature_Granule mature.png")

```

In our workflow, we used the potency method to find the root node (`rootNode_method = "potency"`). Below we show the UMAP plot that ordered the cells based on the root node selected using potency. As expected, the root node is found in the Granule immature cells.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/umap_pseudotime_potency_Granule immature_Granule mature.png")

```

### Identifiation of significant trajectory-variable genes and modules
After that, the pipeline identifies the genes that vary between cell types across the trajectories using the spatial autocorrelation analysis. All the significant trajectory-variable genes were exported in a csv file (`monocleDEG_significant_by_trajectory_Granule immature_Granule mature.csv`).The plot below shows the expression of the top trajectory-variable genes in a feature plot.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/significant_by_trajectory_gene_featurePlot_Granule immature_Granule mature.png")

```

We also visualized the expression levels of top trajectory-variable genes across the pseudotime and colored as cell types.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/significant_by_trajectory_genesInPseudotime_Granule immature_Granule mature.png")

```

The set of the trajectory-variable genes is further grouped into modules using Louvain community analysis. The clustering information is exported in a csv file (`monocleDEG_by_trajectory_moduleInformation_Granule immature_Granule mature.csv`). The module heatmap below shows the expression levels of each module across the two cell types.

```{r, echo = FALSE, eval = TRUE, out.width = "80%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/significant_by_trajectory_moduleHeatmap_Granule immature_Granule mature.png")

```

### Visualization of cell distribution across the pseudotime
The function generate plots that shows the distribution of the cells across the pseudotime. Below the density plot shows the distribution of cells across the pseudotime for the cell types separated by the timepoints.

```{r, echo = FALSE, eval = TRUE, out.width = "60%", out.height = "70%"}

knitr::include_graphics("../scdown/monocle3/images/celltype_distribution_density_Granule immature_Granule mature.png")

```

Below the histogram shows the distribution of cells across the pseudotime for the cell types separated by timepoints.

```{r, echo = FALSE, eval = TRUE, out.width = "60%", out.height = "80%"}

knitr::include_graphics("../scdown/monocle3/images/celltype_distribution_histogram_Granule immature_Granule mature.png")

```


# Output folders and files explained
All outputs of the pipeline will be in the **monocle** folder, which is divided into:

* csv
* images
* figures
* rds

The **csv** subfolder contains:

- full DEG analysis results and from them the significant DEGs per specified model and trajectory.
  
The **images** subfolder contains :

- **cellDistribution**: cell and cell type distribution density plots plus histogram per (subsetted) object and condition.

- **DEG**: 
    
    * significant DEGs found by regression analysis on predefined models (defined in universal variable @DEG_models), visualized in feature plot and violin plot. _This is a general linear regression method, and might work better for models that have continuous values_.
    * significant DEGs found by graph auto-correlation analysis along the trajectory/pseudotime, visualized in feature plot, and plotted their expression level changes along pseudotime.
        * These trajectory-variable genes are subsetted to run DEG analysis with regression again, in an effort to identify genes that are differentially expressed along pseudotime AND between two conditions.

- **pseudotime**: umap by cell types, partitions, learned trajectory, and pseudotime.

The **rds** subfolder contains:

- **rds**: Rds files with the monocle object with the trajectory results 


# Summary

This document outlines a pipeline for analyzing trajectory analysis using monocle3. The pipeline begins by defining key variables, such as the working directory, Seurat object, species, and metadata for cell type annotations. It runs monocle3 on the entire dataset to explore global trajectory analysis and generates visualizations like UMAPs, density, histogram, heatmaps, and violin plots, which highlight trajectories, expression of genes, distribution of cells across pseudotime, and expression of differential expressed genes.

Additionally, the pipeline supports focused analysis by selecting specific cell types or comparing user-defined groups (e.g., age groups 35 and 12). Pairwise comparisons are made to examine differences in trahetcory differences. This pipeline provides a comprehensive tool for exploring and comparing trajectory analysis.

The analysis was performed on a machine with the following specifications:

- **Operating System**: Rocky Linux 8.9 (Green Obsidian)
  - **Product Name**: RHEL
  - **Product Version**: 8.9
- **CPU**: Intel(R) Xeon(R) Platinum 8358 CPU @ 2.60GHz
- **Memory**: 10 GB
